python/ray/train/callbacks/results_preprocessors/aggregate.py

```diff
@@ -0,0 +1,136 @@
+    TRAINING_ITERATION,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class AggregateResultsPreprocessor(ResultsPreprocessor):
+    """A preprocessor that aggregates training metrics from all workers.
+
+    Args:
+        aggregators (Optional[Dict[str, Optional[Tuple|Callable[[List], Any]]]]):
+            A Dict of aggregation methods. The key is the name of the aggregated
+            metric. The value is the aggregation method. If it is a function, the
+            aggregation will be performed by calling this function. If the value
+            is a tuple (var1, var2), var1 specifies which metrics to be aggregated,
+            and var2 is the averaging weights. When var2 is None, the weight will
+            be taken to be equal for all workers. var1 and var2 are expected to be
+            reported using `train.report()`.
+        aggregate_default_metrics (bool): If `aggregate_default_metrics` is True,
+            a list of default metrics will be aggregated. Set it to False otherwise.
+    """
+
+    VALID_SUMMARY_TYPES: Tuple[type] = (
+        int,
+        float,
+        np.float32,
+        np.float64,
+        np.int32,
+        np.int64,
+    )
+    DEFAULT_KEYS: Set[str] = {TIME_THIS_ITER_S, TRAINING_ITERATION}
+
+    def __init__(
+        self,
+        aggregators: Optional[
+            Dict[str, Optional[Union[Tuple, Callable[[List], Any]]]]
+        ] = None,
+        aggregate_default_metrics: bool = True,
+    ):
+        self.aggregators = {} if aggregators is None else aggregators
+        if aggregate_default_metrics:
+            self.aggregators.update(
+                {
+                    "_aggregated" + metrics: (metrics, None)
+                    for metrics in self.DEFAULT_KEYS
+                }
+            )
+
+    def preprocess(self, results: List[Dict] = None) -> List[Dict]:
+        """Average results before sending them to callbacks.
+
+        Args:
+            results List[Dict]: A list of results from all workers. The metrics
+                specified in `metrics_to_average` will be averaged according to
+                their weights. Non-numerical values will be ignored.
+        Returns:
+            A updated list of results.
+        """
+        results = [] if results is None else results
+        if len(results) == 0 or len(self.aggregators) == 0:
+            return results
+
+        reported_metrics = set(results[0].keys())
+        aggregated_metrics = {}
+
+        for name, aggregator in self.aggregators.items():
+
+            if callable(aggregator):
+                aggregated_metrics[name] = aggregator(results)
+            else:
+                aggregated_metrics[name] = self._aggregate_metrics(
+                    *aggregator, results, reported_metrics
+                )
+
+        for result in results:
+            result.update(aggregated_metrics)
+
+        return results
+
+    def _aggregate_metrics(self, metrics, weight, results, reported_metrics):
```

Comment:
`_average_metrics`?

Reply:
Thanks for the suggestion! I will make the update once and for all after hearing back from Matthew and Amog.